### PR TITLE
[WIP] Custom problem and dynamic - (and a use of a dictionnary)

### PR DESCRIPTION
**Pull request by tgouss**

### CODE REVIEW

File: biorbd_optim/problem_type.py [hunk 1]
```diff
@@ -97,7 +107,10 @@ def muscle_activations_and_torque_driven(ocp, nlp):
         nlp["nu"] = nlp["u"].rows()
         nlp["var_controls"]["muscles"] = nlp["nbMuscle"]
 
-        ProblemType.__configure_forward_dyn_func(ocp, nlp, Dynamics.forward_dynamics_torque_muscle_driven)
+        if "dynamic" in nlp["problem_type"]:
+            Problem.configure_forward_dyn_func(ocp, nlp, nlp["problem_type"]["dynamic"])
+        else:
+            Problem.configure_forward_dyn_func(ocp, nlp, Dynamics.forward_dynamics_torque_muscle_driven)
```
Comment (pariterre) [inline biorbd_optim/problem_type.py:113]: Is there a reason to only have this if for this particular condition and not everywhere?

File: examples/getting_started/pendulum.py [hunk 2]
```diff
@@ -13,6 +16,23 @@
 )
 
 
+def custom_dynamic(states, controls, parameters, nlp):
+    Dynamics.apply_parameters(parameters, nlp)
+    q, qdot, tau = Dynamics.dispatch_q_qdot_tau_data(states, controls, nlp)
+
+    qdot_reduced = nlp["q_mapping"].reduce.map(qdot)
+    qddot = biorbd.Model.ForwardDynamics(nlp["model"], q, qdot, tau).to_mx()
+    qddot_reduced = nlp["q_dot_mapping"].reduce.map(qddot)
```
Comment (pariterre) [inline examples/getting_started/pendulum.py:25]: As a matter of simplicity for the user, I would not use reduce mapping in the example. 
(Also, the example should not be for pendulum, but in another file called something like custom_dynamics)

File: examples/getting_started/pendulum.py [hunk 3]
```diff
@@ -13,6 +16,23 @@
 )
 
 
+def custom_dynamic(states, controls, parameters, nlp):
+    Dynamics.apply_parameters(parameters, nlp)
+    q, qdot, tau = Dynamics.dispatch_q_qdot_tau_data(states, controls, nlp)
+
+    qdot_reduced = nlp["q_mapping"].reduce.map(qdot)
+    qddot = biorbd.Model.ForwardDynamics(nlp["model"], q, qdot, tau).to_mx()
```
Comment (pariterre) [inline examples/getting_started/pendulum.py:24]: This line would be shorter if using: 
`qddot = nlp["model"].ForwardDynamics(q, qdot, tau).to_mx()`

File: examples/getting_started/pendulum.py [hunk 4]
```diff
@@ -13,6 +16,23 @@
...
+    qddot_reduced = nlp["q_dot_mapping"].reduce.map(qddot)
+
+    return vertcat(qdot_reduced, qddot_reduced)
```
Comment (pariterre) [inline examples/getting_started/pendulum.py:27]: As a matter of simplicity, I think the vertcat should be done silently inside of the function that called custom_dynamic. The user would therefore be expected to return a list

File: examples/getting_started/pendulum.py [hunk 5]
```diff
@@ -13,6 +16,23 @@
...
+
+
+def custom_torque_driven(ocp, nlp):
+    Problem.configure_q_qdot(nlp, True, False)
```
Comment (pariterre) [inline examples/getting_started/pendulum.py:31]: Use verbose call (not just True, False)

File: examples/getting_started/custom_problem_type_and_dynamics.py [hunk 6]
```diff
@@ -49,10 +49,12 @@ def prepare_ocp(biorbd_model_path, ode_solver=OdeSolver.RK):
...
     # Dynamics
-    problem_type = {"type": ProblemType.CUSTOM, "configure": custom_configure, "dynamic": custom_dynamic}
-
+    if problem_type_custom:
+        problem_type = {"type": ProblemType.CUSTOM, "configure": custom_configure, "dynamic": custom_dynamic}
+    else:
+        problem_type = {"type": ProblemType.TORQUE_DRIVEN, "dynamic": custom_dynamic}
     # problem_type = {"type": ProblemType.TORQUE_DRIVEN, "dynamic": custom_dynamic}  # only custom dynamic
-    # problem_type = {"type": ProblemType.CUSTOM, "function": custom_torque_driven, "dynamic": custom_dynamic}  # custom problem_type and dynamic
+    # problem_type = {"type": ProblemType.CUSTOM, "configure": custom_torque_driven, "dynamic": custom_dynamic}  # custom problem_type and dynamic
```
Comment (pariterre) [inline examples/getting_started/custom_problem_type_and_dynamics.py]: These two line seems useless?

Comment (tgouss) [reply examples/getting_started/custom_problem_type_and_dynamics.py]: They show how to have other configuration than all custom, we can remove the duplicate

Comment (pariterre) [reply examples/getting_started/custom_problem_type_and_dynamics.py]: Aren't they just copy of the above lines?

**Pull request merged**